src/daemon/database/actions.rs

```diff
@@ -591,114 +587,131 @@ pub fn db_mark_activating_vault(db_path: &Path, vault_id: u32) -> Result<(), Dat
+                        transaction.id, db_vault.id
+                    ))
+                })?
+                .try_into()?;
+            let is_fully_signed = db_txs_merge_sigs(&mut transaction, &db_transaction, secp);
+            db_tx.execute(
+                "UPDATE presigned_transactions SET psbt = (?1), fullysigned = (?2) WHERE id = (?3)",
+                params![transaction.psbt.ser(), is_fully_signed, transaction.id],
+            )?;
+        }
 
-        db_tx.execute(
-            "UPDATE presigned_transactions SET psbt = (?1), fullysigned = (?2) WHERE id = (?3)",
-            params![raw_psbt, fully_signed, tx_db_id],
-        )?;
+        Ok(())
+    })
+}
 
-        if fully_signed {
-            // Are there some remaining unsigned revocation txs?
-            if db_tx
-                .prepare(
-                    "SELECT * FROM presigned_transactions WHERE fullysigned = 0 AND type != (?1) AND vault_id = (?2)",
-                )?
-                // All presigned transactions but the Unvault are revocation txs
-                .query(params![TransactionType::Unvault as u32, vault_id])?
-                .next()?
-                .is_none()
-            {
-                // Nope. Mark the vault as 'secured'
-                db_tx
-                    .execute(
-                        "UPDATE vaults SET status = (?1), updated_at = strftime('%s','now') WHERE id = (?2) ",
-                        params![VaultStatus::Secured as u32, vault_id],
-                    )
-                    .map_err(|e| {
-                        DatabaseError(format!("Updating vault to 'secured': {}", e.to_string()))
-                    })?;
-            }
+pub fn db_update_vault_status(db_path: &Path, db_vault: &DbVault) -> Result<(), DatabaseError> {
+    assert!(matches!(
+        db_vault.status,
+        VaultStatus::Unconfirmed
+            | VaultStatus::Funded
+            | VaultStatus::Securing
+            | VaultStatus::Secured
+            | VaultStatus::Activating
+    ));
 
-            // Was it the unvault that was fully signed ? If so, mark the vault as active.
-            if is_unvault {
-                db_tx
-                    .execute(
-                        "UPDATE vaults SET status = (?1), updated_at = strftime('%s','now') WHERE id = (?2) ",
-                        params![VaultStatus::Active as u32, vault_id],
-                    )
-                    .map_err(|e| {
-                        DatabaseError(format!("Updating vault to 'active': {}", e.to_string()))
-                    })?;
+    db_exec(db_path, |db_tx| {
+        let db_transactions: Vec<DbTransaction> = db_tx
+            .prepare("SELECT * FROM presigned_transactions WHERE vault_id = (?1)")?
+            .query_map(params![db_vault.id], |row| row.try_into())?
+            .collect::<rusqlite::Result<Vec<DbTransaction>>>()?;
+
+        let (mut all_signed, mut all_but_unvault_signed) = (true, true);
+        for db_tx in db_transactions {
```

Comment:
This should check the length of the `db_transactions` beforehand or it might set the status to `Active` if the presigned transactions were wiped from the DB. That's a pretty edgy case though, and can be added later. (i'll add a commit to the watchtower integration PR).